katz/src/main/kotlin/katz/Either.kt

```diff
@@ -0,0 +1,194 @@
+     *
+     * Example:
+     * ```
+     * Right(12).map { "flower" } // Result: Right("flower")
+     * Left(12).map { "flower" }  // Result: Left(12)
+     * ```
+     */
+    inline fun <C : Any> map(f: (B) -> C): Either<A, C> =
+            fold({ Left(it) }, { Right(f(it)) })
+
+    /**
+     * Returns `true` if this is a [Right] and its value is equal to `elem` (as determined by `==`),
+     * returns `false` otherwise.
+     *
+     * Example:
+     * ```
+     * Right("something").contains { "something" } // Result: true
+     * Right("something").contains { "anything" }  // Result: false
+     * Left("something").contains { "something" }  // Result: false
+     *  ```
+     *
+     * @param elem    the element to test.
+     * @return `true` if the option has an element that is equal (as determined by `==`) to `elem`, `false` otherwise.
+     */
+    fun <BB, B> contains(elem: () -> BB): Boolean where B : BB =
+            fold({ false }, { it == elem() })
+
+    /**
+     * Returns `false` if [Left] or returns the result of the application of
+     * the given predicate to the [Right] value.
+     *
+     * Example:
+     * ```
+     * Right(12).exists { it > 10 } // Result: true
+     * Right(7).exists { it > 10 }  // Result: false
+     *
+     * val left: Either<Int, Int> = Left(12)
+     * left.exists { it > 10 }      // Result: false
+     * ```
+     */
+    inline fun exists(predicate: (B) -> Boolean): Boolean =
+            fold({ false }, { predicate(it) })
+
+    /**
+     * Returns a [Option.Some] containing the [Right] value
+     * if it exists or a [Option.None] if this is a [Left].
+     *
+     * Example:
+     * ```
+     * Right(12).toOption() // Result: Some(12)
+     * Left(12).toOption()  // Result: None
+     * ```
+     */
+    fun toOption(): Option<B> =
+            fold({ Option.None }, { Option.Some(it) })
+
+    /**
+     * The left side of the disjoint union, as opposed to the [Right] side.
+     */
+    data class Left<out A : Any>(val a: A) : Either<A, Nothing>() {
+        override val isLeft = true
+        override val isRight = false
+    }
+
+    /**
+     * The right side of the disjoint union, as opposed to the [Left] side.
+     */
+    data class Right<out B : Any>(val b: B) : Either<Nothing, B>() {
+        override val isLeft = true
+        override val isRight = false
+    }
+}
+
+/**
+ * Binds the given function across [Either.Right].
+ *
+ * @param f The function to bind across [Either.Right].
+ */
+inline fun <A : Any, B : Any, C : Any> Either<A, B>.flatMap(f: (B) -> Either<A, C>): Either<A, C> =
+        fold({ Either.Left(it) }, { f(it) })
+
+/**
+ * Returns the value from this [Either.Right] or the given argument if this is a [Either.Left].
+ *
+ * Example:
+ * ```
+ * Right(12).getOrElse(17) // Result: 12
+ * Left(12).getOrElse(17)  // Result: 17
+ * ```
+ */
+inline fun <B : Any> Either<*, B>.getOrElse(default: () -> B): B =
+        fold({ default() }, { b -> b })
```

Comment:
I did it the same, but now I think that it's more clear in this way:

`+inline fun <B : Any> Either<*, B>.getOrElse(default: () -> B): B = fold({ default() }, { it })`

`it` better than `b -> b`